src/ast.h

```diff
@@ -0,0 +1,299 @@
+/*
+ * (ast.h | 8 Mar 19 | Amr Anwar)
+ * 
+ * The abstract syntax trees structs
+ * 
+*/
+
+#ifndef ast_h
+#define ast_h
+
+#include <stdint.h>
+
+#include "list.h"
+#include "lexer.h"
+
+typedef struct expr expr;
+typedef struct pattern pattern;
+
+typedef struct {
+    list *stmts;
+} piece;
+
+typedef enum {
+    let_stmt_type,
+    fn_stmt_type,
+    ret_stmt_type,
+    expr_stmt_type,
+    fixed_stmt_type
+} stmt_type;
+
+typedef enum {
+    BREAK,
+    CONTINUE,
+} fixed_stmt;
+
+typedef struct {
+    char* ident;
+    list* patts;
+} param_list;
+
+typedef struct {
+    char *n_ident;
+    list *param_list;
+    piece *fn_p;
+} fn_stmt;
+
+typedef struct {
+    char *ident;
+    expr *expression;
+} let_stmt;
+
+typedef struct {
+    expr *expression;
+} ret_stmt;
+
+typedef struct {
+    expr *expression;
+} expr_stmt;
+
+typedef struct {
+    stmt_type type;
+    union {
+        let_stmt * ls;
+        fn_stmt *fns;
+        ret_stmt *rs;
+        expr_stmt *es;
+        fixed_stmt fs;
+    } obj;
+} stmt;
+
+typedef enum {
+    lit_expr_type,
+    prefix_expr_type,
+    infix_expr_type,
+    index_expr_type,
+    access_expr_type,
+    group_expr_type,
+    call_expr_type,
+    if_expr_type,
+    for_expr_type,
+    while_expr_type,
+    match_expr_type,
+    ident_expr_type,
+} expr_type;
+
+typedef struct {
+    expr *expression;
+    char *n_ident;
+} access_expr;
+
+typedef struct {
+    token_type op;
+    expr *left;
+    expr *right;
+} infix_expr;
+
+typedef struct {
+    token_type op;
+    expr *expression;
+} prefix_expr;
+
+typedef struct {
+    expr *outer_e;
+    expr *inner_e;
+} index_expr;
+
+typedef struct {
+    expr *group_e;
+} group_expr;
+
+typedef struct {
+    expr *expression;
+    list *exprs;
+} expr_list;
+
+typedef struct {
+    expr *expr;
+    list *expression_l;
+} call_expr;
+
+typedef struct {
+    expr *if_e;
+    piece *if_p;
+    expr *elif_e; 
+    piece *elif_p;
+    piece *else_p;
+} if_expr;
+
+typedef struct {
+    char *name_i;
+    expr *for_e;
+    piece *for_p;
+} for_expr;
+
+typedef struct {
+    expr *while_e;
+    piece *while_p;
+} while_expr;
+
+typedef struct {
+    expr *match_e;
+    list *match_bs; 
+} match_expr;
+
+typedef enum {
+    fn_lit_type,
+    list_lit_type,
+    record_lit_type,
+    int_lit_type,
+    float_lit_type,
+    str_lit_type,
+    rstr_lit_type,
+    true_type,
+    false_type,
+    nil_lit_type,
+} lit_type;
+
+typedef struct {
+    int64_t i;
+} int_lit;
+
+typedef struct {
+    double f;
+} float_lit;
+
+typedef struct {
+    char *s;
+} str_lit;
+
+typedef struct {
+    char *s;
+} rstr_lit;
+
+typedef struct {
+    list *fn_param_l;
+    piece *fn_p;
+} fn_lit;
+
+typedef struct {
+    expr *list_e;
+    list *list_exprs;
+} list_lit;
+
+typedef struct {
+    char *name;
+    expr *experssion;
+} record_field;
+
+typedef struct {
+    record_field *record_f;
+    list *record_fields;
+} record_lit;
```

Comment:
another unnecessary wrapping.
you should use a list of record_fields. and don't call them `record_fields` that the name of the type itself.
```c
typdef struct {
    list *fields;
}
```
you could discard the  `record field` totally and use the `char*` and `expr` lists directly
(i think that's a better approach)
```c
typedef struct {
    list *names;
    list *values;
} record_lit;
```